rmw_zenoh_cpp/src/rmw_zenoh.cpp

```diff
@@ -1672,11 +1998,67 @@ rmw_take_response(
   void * ros_response,
   bool * taken)
 {
-  static_cast<void>(client);
-  static_cast<void>(request_header);
-  static_cast<void>(ros_response);
-  static_cast<void>(taken);
-  return RMW_RET_UNSUPPORTED;
+  *taken = false;
+  RCUTILS_LOG_INFO_NAMED("rmw_zenoh_cpp", "[rmw_take_response]");
+
+  RMW_CHECK_ARGUMENT_FOR_NULL(client, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(ros_response, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(taken, RMW_RET_INVALID_ARGUMENT);
+
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    client,
+    client->implementation_identifier,
+    rmw_zenoh_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+
+  RMW_CHECK_FOR_NULL_WITH_MSG(
+    client->service_name, "client has no service name", RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_FOR_NULL_WITH_MSG(
+    client->data, "client implementation pointer is null", RMW_RET_INVALID_ARGUMENT);
+
+
+  auto client_data = static_cast<rmw_client_data_t *>(client->data);
+
+  std::unique_ptr<saved_msg_data> msg_data = nullptr;
+
+  {
+    std::lock_guard<std::mutex> lock(client_data->message_mutex);
+    if (client_data->message == nullptr) {
+      // TODO(francocipollone): Verify behavior.
+      RCUTILS_LOG_ERROR_NAMED("rmw_zenoh_cpp", "[rmw_take_response] Response message is empty");
+      return RMW_RET_ERROR;
+    }
+
+    msg_data = std::move(client_data->message);
+    client_data->message.release();
```

Comment:
While this releases the unique_ptr, does it free the sample we took ownership of using `zc_sample_payload_rcinc()`? Anyways if we switch to storing the owned replies directly, we will not have this problem.